App/Services/HostnameSuffixGetter.cs

```diff
@@ -0,0 +1,145 @@
+    }
+
+    ~HostnameSuffixGetter()
+    {
+        _credentialManager.CredentialsChanged -= HandleCredentialsChanged;
+    }
+
+    private void HandleCredentialsChanged(object? sender, CredentialModel credentials)
+    {
+        using var _ = _lock.Lock();
+        _logger.LogDebug("credentials updated with state {state}", credentials.State);
+        _credentialModel = credentials;
+        if (credentials.State != CredentialState.Valid) return;
+
+        _dirty = true;
+        if (!_getInProgress)
+        {
+            _getInProgress = true;
+            Task.Run(Refresh).ContinueWith(MaybeRefreshAgain);
+        }
+    }
+
+    private async Task Refresh()
+    {
+        _logger.LogDebug("refreshing domain suffix");
+        CredentialModel credentials;
+        using (_ = await _lock.LockAsync())
+        {
+            credentials = _credentialModel;
+            if (credentials.State != CredentialState.Valid)
+            {
+                _logger.LogDebug("abandoning refresh because credentials are now invalid");
+                return;
+            }
+
+            _dirty = false;
+        }
+
+        var client = _clientFactory.Create(credentials.CoderUrl!.ToString());
+        client.SetSessionToken(credentials.ApiToken!);
+        using var timeoutSrc = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+        var connInfo = await client.GetAgentConnectionInfoGeneric(timeoutSrc.Token);
+
+        // older versions of Coder might not set this
+        var suffix = string.IsNullOrEmpty(connInfo.HostnameSuffix)
+            ? DefaultSuffix
+            // and, it doesn't include the leading dot.
+            : "." + connInfo.HostnameSuffix;
+
+        var changed = false;
+        using (_ = await _lock.LockAsync(CancellationToken.None))
+        {
+            if (_domainSuffix != suffix) changed = true;
+            _domainSuffix = suffix;
+        }
+
+        if (changed)
+        {
+            _logger.LogInformation("got new domain suffix '{suffix}'", suffix);
+            // grab a local copy of the EventHandler to avoid TOCTOU race on the `?.` null-check
+            var del = SuffixChanged;
+            del?.Invoke(this, suffix);
+        }
+        else
+        {
+            _logger.LogDebug("domain suffix unchanged '{suffix}'", suffix);
+        }
+    }
+
+    private async Task MaybeRefreshAgain(Task prev)
+    {
+        if (prev.IsFaulted)
+        {
+            _logger.LogError(prev.Exception, "failed to query domain suffix");
+            // back off here before retrying. We're just going to use a fixed, long
+            // delay since this just affects UI stuff; we're not in a huge rush as
+            // long as we eventually get the right value.
+            await Task.Delay(TimeSpan.FromSeconds(10));
+        }
+
+        using var l = await _lock.LockAsync(CancellationToken.None);
+        if ((_dirty || prev.IsFaulted) && _credentialModel.State == CredentialState.Valid)
```

Comment:
`_dirty` seems to only be true here if the credentials became invalid between when the event was handled and the task started. So the only way for this to trigger would be if credentials became valid, then became invalid, then became valid again, which seems very unlikely to actually happen. The user would need to sign in or start the app, then sign out, then enter their credentials and sign back in (with API auth check latency included).

You could probably also accomplish the same thing without needing a bool by throwing in the credential state guard at the top of `Refresh()`, and only relying on `prev.IsFaulted` here

Reply:
The idea of `_dirty` is that while we are making the network call a new, valid credential for a different deployment could come in. That is, just because we had a valid credential when we started our work, and have a valid credential at the end, it doesn't mean that _it was the same credential_. If that happens then we refresh to ensure we have the correct value from the new credential.